File: src/shared/session.ts
export const COOKIE = 'sidepad_session';

function secret() {
  const value = String(process.env.SESSION_SECRET || '').trim();
  if (!value) throw new Error('Set SESSION_SECRET');
  return value;
}

function toHex(buffer: ArrayBuffer): string {
  return [...new Uint8Array(buffer)]
    .map((byte) => byte.toString(16).padStart(2, '0'))
    .join('');
}

async function hmacHex(message: string): Promise<string> {
  const key = await crypto.subtle.importKey(
    'raw',
    new TextEncoder().encode(secret()),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  );
  const signature = await crypto.subtle.sign(
    'HMAC',
    key,
    new TextEncoder().encode(message)
  );
  return toHex(signature);
}

function safeEqual(left: string, right: string): boolean {
  if (left.length !== right.length) return false;

  let isEqual = true;

  for (let index = 0; index < left.length; index++) {
    if (left[index] !== right[index]) isEqual = false;
  }

  return isEqual;
}

export async function signSession(userId: string): Promise<string> {
  const signature = await hmacHex(userId);
  return `${userId}.${signature}`;
}

export async function readSession(
  token: string | undefined | null
): Promise<string | null> {
  if (!token || !token.includes('.')) return null;

  const separatorIndex = token.indexOf('.');
  const userId = token.slice(0, separatorIndex);
  const signature = token.slice(separatorIndex + 1);

  if (!userId || !signature) return null;

  try {
    const expected = await hmacHex(userId);
    if (!safeEqual(signature, expected)) return null;
    return userId;
  } catch {
    return null;
  }
}

export async function isValidSession(
  token: string | undefined | null
): Promise<boolean> {
  return (await readSession(token)) !== null;
}

File: backend/app/shared/session.py
import hashlib
import hmac
import os

COOKIE = "sidepad_session"


def _secret() -> str:
    value = str(os.environ.get("SESSION_SECRET") or "").strip()
    if not value:
        raise RuntimeError("Set SESSION_SECRET")
    return value


def _hmac_hex(message: str) -> str:
    return hmac.new(
        _secret().encode("utf-8"),
        message.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()


def _safe_equal(left: str, right: str) -> bool:
    if len(left) != len(right):
        return False
    return hmac.compare_digest(left, right)


def sign_session(user_id: str) -> str:
    return f"{user_id}.{_hmac_hex(user_id)}"


def read_session(token: str | None) -> str | None:
    if not token or "." not in token:
        return None

    separator_index = token.index(".")
    user_id = token[:separator_index]
    signature = token[separator_index + 1 :]

    if not user_id or not signature:
        return None

    try:
        expected = _hmac_hex(user_id)
        if not _safe_equal(signature, expected):
            return None
        return user_id
    except Exception:
        return None
